### PR TITLE
[tuner]: use translation_info binding

### PR DESCRIPTION
**Pull request by bangtianliu**
This PR is relevant to the task in https://github.com/nod-ai/shark-ai/issues/453 : use IREE bindings for compilation info (incl., lowering_config and translation_info).

Use translation_info from IREE  python binding.

### CODE REVIEW

File: tuner/tuner/candidate_gen_test.py [hunk 1]
```diff
@@ -56,13 +57,17 @@ def test_apply_params_mmt(tuner_ctx: common.TunerContext) -> None:
         subgroup_m_count=16,
         subgroup_n_count=16,
     )
+    pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+        iree_codegen.DispatchLoweringPassPipeline.None_
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:61]: The pipeline should be llvmgpuvectorize. Also everywhere else

**Review by kuhar — CHANGES_REQUESTED** [review]

File: tuner/tuner/candidate_gen_test.py [hunk 2]
```diff
@@ -56,13 +57,17 @@ def test_apply_params_mmt(tuner_ctx: common.TunerContext) -> None:
...
+    )
+    pipeline_option = iree_gpu.PipelineOptionsAttr.get(prefetch_shared_memory=True)
+    pipeline_option_dict = ir.DictAttr.get({"gpu_pipeline_options": pipeline_option})
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:64]: Can you put the gpu_pipeline_options key name as a constant somewhere in `common.py`?

File: tuner/tuner/candidate_gen.py [hunk 3]
```diff
@@ -147,7 +150,7 @@ def get_transform_function_mmt(
     %config = transform.param.constant #iree_codegen.compilation_info<
         lowering_config = {configuration.lowering_config}>,
         translation_info = #iree_codegen.translation_info<LLVMGPUVectorDistribute
-        workgroup_size = [{wg_x}, {wg_y}, {wg_z}] subgroup_size = {configuration.subgroup_size},
+        workgroup_size = [{wg_x}, {wg_y}, {wg_z}] subgroup_size = {configuration.translation_info.subgroup_size},
         {{mma_schedule = #iree_gpu.mma_schedule<
             intrinsic = {intrinsic},
             subgroup_m_count = {subgroup_m_count}, subgroup_n_count = {subgroup_n_count}>
```
Comment (kuhar) [inline tuner/tuner/candidate_gen.py:156]: Can we print the whole translation info attr here?

File: tuner/tuner/candidate_gen.py [hunk 4]
```diff
@@ -219,7 +222,7 @@ def get_transform_function_conv(
         %config = transform.param.constant #iree_codegen.compilation_info<
         lowering_config = {configuration.lowering_config}>,
         translation_info = #iree_codegen.translation_info<LLVMGPUVectorDistribute
-        workgroup_size = [{wg_x}, {wg_y}, {wg_z}] subgroup_size = {configuration.subgroup_size},
+        workgroup_size = [{wg_x}, {wg_y}, {wg_z}] subgroup_size = {configuration.translation_info.subgroup_size},
```
Comment (kuhar) [inline tuner/tuner/candidate_gen.py:225]: Also here

File: tuner/tuner/candidate_gen.py [hunk 5]
```diff
@@ -286,7 +289,7 @@ def get_transform_function_broadcast_rhs_mmt(
 %config = transform.param.constant #iree_codegen.compilation_info<
     lowering_config = {configuration.lowering_config}>,
     translation_info = #iree_codegen.translation_info<LLVMGPUVectorDistribute
-    workgroup_size = [{wg_x}, {wg_y}, {wg_z}] subgroup_size = {configuration.subgroup_size},
+    workgroup_size = [{wg_x}, {wg_y}, {wg_z}] subgroup_size = {configuration.translation_info.subgroup_size},
```
Comment (kuhar) [inline tuner/tuner/candidate_gen.py:292]: Also here

File: tuner/tuner/candidate_gen.py [hunk 6]
```diff
@@ -371,7 +374,7 @@ def get_transform_function_batch_mmt(
...
```
Comment (kuhar) [inline tuner/tuner/candidate_gen.py:377]: Also here

File: tuner/tuner/candidate_gen.py [hunk 7]
```diff
@@ -443,7 +446,7 @@ def get_transform_function_batch_matmul(
         %config = transform.param.constant #iree_codegen.compilation_info<
         lowering_config = {configuration.lowering_config}>,
         translation_info = #iree_codegen.translation_info<LLVMGPUPadAndVectorDistribute
-        workgroup_size = [{wg_x}, {wg_y}, {wg_z}] subgroup_size = {configuration.subgroup_size},
+        workgroup_size = [{wg_x}, {wg_y}, {wg_z}] subgroup_size = {configuration.translation_info.subgroup_size},
```
Comment (kuhar) [inline tuner/tuner/candidate_gen.py:449]: Also here

File: tuner/tuner/candidate_gen_test.py [hunk 8]
```diff
@@ -56,13 +57,19 @@ def test_apply_params_mmt(tuner_ctx: common.TunerContext) -> None:
         subgroup_m_count=16,
         subgroup_n_count=16,
     )
+    pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+        iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorize
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:61]: this is not the right pipeline

File: tuner/tuner/candidate_gen_test.py [hunk 9]
```diff
@@ -118,15 +125,23 @@ def test_apply_params_conv(tuner_ctx: common.TunerContext) -> None:
         subgroup_m_count=1,
         subgroup_n_count=4,
     )
+    pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+        iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorize
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:129]: also here

File: tuner/tuner/candidate_gen_test.py [hunk 10]
```diff
@@ -191,11 +206,19 @@ def test_apply_params_contract(tuner_ctx: common.TunerContext) -> None:
...
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:210]: also here

File: tuner/tuner/candidate_gen_test.py [hunk 11]
```diff
@@ -246,11 +269,19 @@ def test_apply_params_batch_matmul(tuner_ctx: common.TunerContext) -> None:
         subgroup_m_count=2,
         subgroup_n_count=2,
     )
+    pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+        iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorize
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:273]: also here

File: tuner/tuner/candidate_gen_test.py [hunk 12]
```diff
@@ -304,11 +335,19 @@ def test_apply_params_batch_mmt_float(tuner_ctx: common.TunerContext) -> None:
...
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:339]: also here

File: tuner/tuner/candidate_gen_test.py [hunk 13]
```diff
@@ -360,11 +399,19 @@ def test_apply_params_batch_mmt_int(tuner_ctx: common.TunerContext) -> None:
...
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:403]: also here

File: tuner/tuner/candidate_gen_test.py [hunk 14]
```diff
@@ -440,11 +487,19 @@ def test_apply_params_broadcast_rhs_mmt(tuner_ctx: common.TunerContext) -> None:
...
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:491]: also here

File: tuner/tuner/common.py [hunk 15]
```diff
@@ -112,13 +113,15 @@ def is_comptible(mma_intrinsic: iree_gpu.MMAIntrinsic) -> bool:
...
-    subgroup_size: int
-    workgroup_size: list[int]
+    translation_info: iree_codegen.TranslationInfoAttr
     lowering_config: iree_gpu.LoweringConfigAttr
-    gpu_pipeline_options: iree_gpu.PipelineOptionsAttr
     waves_per_eu: int
 
 
+# The key name for GPUPipelineOptionsAttr in the translation info config dictionary.
+GPU_PIPELINE_OPTIONS = "gpu_pipeline_options"
```
Comment (kuhar) [inline tuner/tuner/common.py:122]:
```suggestion
GPU_PIPELINE_OPTIONS_KEY = "gpu_pipeline_options"
```

File: tuner/tuner/common_test.py [hunk 16]
```diff
@@ -84,11 +85,19 @@ def test_get_pipeline_config(tuner_ctx: common.TunerContext) -> None:
         subgroup_m_count=1,
         subgroup_n_count=1,
     )
+    pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+        iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorize
```
Comment (kuhar) [inline tuner/tuner/common_test.py:89]: wrong pipeline

File: tuner/tuner/common_test.py [hunk 17]
```diff
@@ -207,11 +225,19 @@ def test_get_lowering_config(tuner_ctx: common.TunerContext) -> None:
         == "#iree_gpu.lowering_config<{reduction = [0, 0, 16], subgroup_m_count = 1 : i64, subgroup_n_count = 1 : i64, workgroup = [4, 8, 0]}>"
     )
 
+    pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+        iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorize
```
Comment (kuhar) [inline tuner/tuner/common_test.py:229]: wrong pipeline

File: tuner/tuner/dispatch_constraints.py [hunk 18]
```diff
@@ -244,13 +244,20 @@ def generate_solutions(
             subgroup_m_count=lookup(sg_m_cnt),
             subgroup_n_count=lookup(sg_n_cnt),
         )
-        config = Configuration(
-            lookup(subgroup_size),
+        pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+            iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorDistribute
```
Comment (kuhar) [inline tuner/tuner/dispatch_constraints.py:248]: this one is correct

File: tuner/tuner/dispatch_parser_test.py [hunk 19]
```diff
@@ -50,11 +51,19 @@ def test_get_mmt_tile_sizes(tuner_ctx: common.TunerContext) -> None:
         subgroup_m_count=1,
         subgroup_n_count=4,
     )
-    config = dispatch_parser.Configuration(
-        subgroup_size=0,
-        workgroup_size=[],
+    pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+        iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorize
```
Comment (kuhar) [inline tuner/tuner/dispatch_parser_test.py:55]: wrong pipeline

File: tuner/tuner/dispatch_parser_test.py [hunk 20]
```diff
@@ -73,11 +82,19 @@ def test_get_conv_tile_sizes(tuner_ctx: common.TunerContext) -> None:
         subgroup_m_count=1,
         subgroup_n_count=4,
     )
-    config = dispatch_parser.Configuration(
-        subgroup_size=64,
-        workgroup_size=[256, 1, 1],
+    pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+        iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorize
```
Comment (kuhar) [inline tuner/tuner/dispatch_parser_test.py:86]: wrong pipeline

File: tuner/tuner/dispatch_parser_test.py [hunk 21]
```diff
@@ -95,11 +112,19 @@ def test_get_contract_tile_sizes(tuner_ctx: common.TunerContext) -> None:
         subgroup_m_count=1,
         subgroup_n_count=1,
     )
-    config = dispatch_parser.Configuration(
-        subgroup_size=32,
-        workgroup_size=[16, 16, 1],
+    pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+        iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorize
```
Comment (kuhar) [inline tuner/tuner/dispatch_parser_test.py:116]: wrong pipeline

File: tuner/tuner/candidate_gen_test.py [hunk 22]
```diff
@@ -56,13 +57,19 @@ def test_apply_params_mmt(tuner_ctx: common.TunerContext) -> None:
         subgroup_m_count=16,
         subgroup_n_count=16,
     )
+    pipeline_attr = iree_codegen.DispatchLoweringPassPipelineAttr.get(
+        iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorize
+    )
+    pipeline_option = iree_gpu.PipelineOptionsAttr.get(prefetch_shared_memory=True)
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:63]:
```suggestion
    pipeline_options = iree_gpu.PipelineOptionsAttr.get(prefetch_shared_memory=True)
```

File: tuner/tuner/candidate_gen_test.py [hunk 23]
```diff
@@ -56,13 +57,19 @@ def test_apply_params_mmt(tuner_ctx: common.TunerContext) -> None:
...
+    pipeline_option_dict = ir.DictAttr.get(
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:64]:
```suggestion
    pipeline_options_dict = ir.DictAttr.get(
```

File: tuner/tuner/candidate_gen_test.py (hunk 22 above)
Comment (kuhar) [reply tuner/tuner/candidate_gen_test.py:63]: also everywhere else

File: tuner/tuner/candidate_gen_test.py (hunk 23 above)
Comment (kuhar) [reply tuner/tuner/candidate_gen_test.py:64]: also everywhere else

File: tuner/tuner/candidate_gen.py [hunk 24]
```diff
@@ -369,13 +321,8 @@ def get_transform_function_batch_mmt(
...
 %config = transform.param.constant #iree_codegen.compilation_info<
-    lowering_config = {configuration.lowering_config}>,
-    translation_info = #iree_codegen.translation_info<LLVMGPUVectorDistribute
-    workgroup_size = [{wg_x}, {wg_y}, {wg_z}] subgroup_size = {configuration.subgroup_size},
-    {{mma_schedule = #iree_gpu.mma_schedule<
-        intrinsic = {intrinsic},
-        subgroup_m_count = {subgroup_m_count}, subgroup_n_count = {subgroup_n_count}>
-    {extra_config}}}>
+    lowering_config = {configuration.lowering_config},
+    translation_info ={configuration.translation_info}
```
Comment (kuhar) [inline tuner/tuner/candidate_gen.py:325]:
```suggestion
    translation_info = {configuration.translation_info}
```

File: tuner/tuner/candidate_gen_test.py [hunk 25]
```diff
@@ -56,14 +57,23 @@ def test_apply_params_mmt(tuner_ctx: common.TunerContext) -> None:
...
+        iree_codegen.DispatchLoweringPassPipeline.LLVMGPUVectorDistribute
+    )
+    pipeline_options = iree_gpu.PipelineOptionsAttr.get(prefetch_shared_memory=True)
+    waves_per_eu_dict = ir.DictAttr.get({"amdgpu-waves-per-eu": ir.StringAttr.get("8")})
+    config_dict = ir.DictAttr.get(
+        {
+            common.GPU_PIPELINE_OPTIONS_KEY: pipeline_options,
+            common.LLVM_FUNC_ATTRS_KEY: waves_per_eu_dict,
+        }
+    )
```
Comment (kuhar) [inline tuner/tuner/candidate_gen_test.py:70]: You can create a helper function to build config_dict for you, similar to the helper function `get_lowering_config`

Comment (bangtianliu) [reply tuner/tuner/candidate_gen_test.py:70]: Good suggestion! will do it.

File: tuner/tuner/common.py [hunk 26]
```diff
@@ -157,15 +163,29 @@ def get_lowering_config(
...
-    extra_config = ""
-    pipeline_options = configuration.gpu_pipeline_options
-    if pipeline_options != iree_gpu.PipelineOptionsAttr.get():
-        extra_config += f", gpu_pipeline_options = {pipeline_options}"
-
-    if configuration.waves_per_eu != 2:
-        extra_config += f', llvm_func_attrs = {{"amdgpu-waves-per-eu" = "{configuration.waves_per_eu}"}}'
-    return extra_config
+def get_translation_info_config(
+    pipeline_options: iree_gpu.PipelineOptionsAttr, waves_per_eu: int | str
```
Comment (kuhar) [inline tuner/tuner/common.py:167]: I think it would be easier to always require `waves_per_eu` to be an int

File: tuner/tuner/common.py [hunk 27]
```diff
@@ -157,15 +163,29 @@ def get_lowering_config(
...
-def get_pipeline_config(configuration: Configuration) -> str:
-    extra_config = ""
-    pipeline_options = configuration.gpu_pipeline_options
-    if pipeline_options != iree_gpu.PipelineOptionsAttr.get():
-        extra_config += f", gpu_pipeline_options = {pipeline_options}"
-
-    if configuration.waves_per_eu != 2:
-        extra_config += f', llvm_func_attrs = {{"amdgpu-waves-per-eu" = "{configuration.waves_per_eu}"}}'
-    return extra_config
+def get_translation_info_config(
```
Comment (kuhar) [inline tuner/tuner/common.py:166]: Could you add a comment explaining what the purpose of this function is? Maybe show a piece of IR and where this attribute fits?

File: tuner/tuner/common.py [hunk 28]
```diff
@@ -157,15 +163,45 @@ def get_lowering_config(
...
-def get_pipeline_config(configuration: Configuration) -> str:
-    extra_config = ""
-    pipeline_options = configuration.gpu_pipeline_options
-    if pipeline_options != iree_gpu.PipelineOptionsAttr.get():
-        extra_config += f", gpu_pipeline_options = {pipeline_options}"
-
-    if configuration.waves_per_eu != 2:
-        extra_config += f', llvm_func_attrs = {{"amdgpu-waves-per-eu" = "{configuration.waves_per_eu}"}}'
-    return extra_config
+# Generate a config dictionary in translation info
```
Comment (kuhar) [inline tuner/tuner/common.py:166]: This doesn't generate a dictionary **in** translation info. I'd say something like this instead:
```suggestion
# Generate a config dictionary used in the translation_info attribute.
```

File: tuner/tuner/common.py [hunk 29]
```diff
@@ -157,15 +163,45 @@ def get_lowering_config(
...
+                    }
+                >
+    Example Usage:
+            pipeline_options = iree_gpu.PipelineOptionsAttr.get(...)
+            waves_per_eu = 3
+
+            config_dict = get_translation_info_config(
+                pipeline_options=pipeline_options,
+                waves_per_eu=waves_per_eu
+            )
```
Comment (kuhar) [inline tuner/tuner/common.py:187]: I don't think we need this, the usage pattern is pretty obvious

**Review by kuhar — APPROVED** [review]
LGTM % comments

File: tuner/tuner/common.py [hunk 30]
```diff
@@ -177,14 +177,6 @@ def get_translation_info_config(
...
-    Example Usage:
-            pipeline_options = iree_gpu.PipelineOptionsAttr.get(...)
-            waves_per_eu = 3
-
-            config_dict = get_translation_info_config(
-                pipeline_options=pipeline_options,
-                waves_per_eu=waves_per_eu
-            )
 
     this 'config_dict' is subsequently used afterward to generate the 'translation_info' in the above example IR."""
```
Comment (kuhar) [inline tuner/tuner/common.py:181]: Also this is now outdated

File: tuner/tuner/common.py [hunk 31]
```diff
@@ -157,15 +163,36 @@ def get_lowering_config(
...
+def get_translation_info_config(
+    pipeline_options: iree_gpu.PipelineOptionsAttr, waves_per_eu: int
+) -> ir.DictAttr:
+    """
+    Example IR
+    translation_info = #iree_codegen.translation_info<
+                    pipeline = LLVMGPUVectorDistribute
+                    workgroup_size = [512, 1, 1]
+                    subgroup_size = 64,
+                    {gpu_pipeline_options = #iree_gpu.pipeline_options<>
```
Comment (kuhar) [inline tuner/tuner/common.py:176]:
```suggestion
                    {gpu_pipeline_options = #iree_gpu.pipeline_options<...>,
```

File: tuner/tuner/common.py [hunk 32]
```diff
@@ -157,15 +163,36 @@ def get_lowering_config(
...
+# Generate a config dictionary used in translation_info attribute.
+def get_translation_info_config(
+    pipeline_options: iree_gpu.PipelineOptionsAttr, waves_per_eu: int
+) -> ir.DictAttr:
+    """
+    Example IR
+    translation_info = #iree_codegen.translation_info<
+                    pipeline = LLVMGPUVectorDistribute
+                    workgroup_size = [512, 1, 1]
+                    subgroup_size = 64,
```
Comment (kuhar) [inline tuner/tuner/common.py:175]:
```suggestion
                    pipeline = LLVMGPUVectorDistribute workgroup_size = [512, 1, 1] subgroup_size = 64,
```

File: tuner/tuner/common.py [hunk 33]
```diff
@@ -157,15 +163,34 @@ def get_lowering_config(
...
-    return extra_config
+# Generate a config dictionary used in translation_info attribute.
+def get_translation_info_config(
+    pipeline_options: iree_gpu.PipelineOptionsAttr, waves_per_eu: int
+) -> ir.DictAttr:
+    """
+    Example IR
+    translation_info = #iree_codegen.translation_info<
+                    pipeline = LLVMGPUVectorDistribute workgroup_size = [512, 1, 1] subgroup_size = 64,
+                    {gpu_pipeline_options = #iree_gpu.pipeline_options<...>
```
Comment (kuhar) [inline tuner/tuner/common.py:174]:
```suggestion
                    {gpu_pipeline_options = #iree_gpu.pipeline_options<...>,
```

**Pull request merged**